In [28]:
%pip install folium pandas OSMnx networkx scikit-learn matplotlib numpy

Note: you may need to restart the kernel to use updated packages.


In [29]:
import solution_reader
import os
import pandas as pd
import numpy as np

In [30]:
# Files

solution_directory = "./../output/selection"
best_known_file = "./../best_known/best_known_LL01_100.txt"

# Best known dataframe
best_known_df = solution_reader.read_best_known(best_known_file)

# Solution dataframe
solutions_list = []
files = os.listdir(solution_directory)
for solution_file in files :
    solution = solution_reader.PDPTWCompleteSolution(solution_directory + "/" + solution_file)
    
    solution_dict = {
    "Instance": solution.get_instance(),
    "Vehicles": solution.get_vehicles(),
    "Distance": solution.get_cost(),
    "Reference": solution.get_reference(),
    "Date": solution.get_date()
    }

    solutions_list.append(solution_dict)

solutions_df = pd.DataFrame(solutions_list)

best_known_df = best_known_df[best_known_df['Instance'].isin(solutions_df['Instance'])]
solutions_df.sort_values(by=["Instance"], inplace=True)
best_known_df.sort_values(by=["Instance"], inplace=True)

solutions_df.reset_index(drop=True, inplace=True)
best_known_df.reset_index(drop=True, inplace=True)

print(best_known_df)
print(solutions_df)

   Instance  Vehicles  Distance Reference       Date
0     lc101        10    828.94    Li&Lim       2001
1     lc102        10    828.94    Li&Lim       2001
2     lc103         9   1035.35       BVH  27-jun-03
3     lc201         3    591.56    Li&Lim       2001
4     lc202         3    591.56    Li&Lim       2001
5     lc203         3    591.17  SAM::OPT  11-mar-03
6     lr101        19   1650.80    Li&Lim       2001
7     lr102        17   1487.57    Li&Lim       2001
8     lr103        13   1292.68    Li&Lim       2001
9     lr201         4   1253.23  SAM::OPT  28-feb-03
10    lr202         3   1197.67    Li&Lim       2001
11    lr203         3    949.40    Li&Lim       2001
12   lrc101        14   1708.80    Li&Lim       2001
13   lrc102        12   1558.07  SAM::OPT  19-feb-03
14   lrc103        11   1258.74    Li&Lim       2001
15   lrc201         4   1406.94  SAM::OPT  28-feb-03
16   lrc202         3   1374.27    Li&Lim       2001
17   lrc203         3   1089.07    Li&Lim     

/home/a24jacqb/Documents/Code/pdptw-main/python/solution_reader.py:14: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df = pd.read_csv(file, delim_whitespace=True, names=["Instance", "Vehicles", "Distance", "Reference", "Date"], skiprows=1)


In [31]:
sum_vehicles_best_known = np.sum(best_known_df["Vehicles"])
sum_vehicles_solution = np.sum(solutions_df["Vehicles"])

sum_distance_best_known = np.sum(best_known_df["Distance"])
sum_distance_solution = np.sum(solutions_df["Distance"])

mean_vehicles_best_known = np.mean(best_known_df["Vehicles"])
mean_vehicles_solution = np.mean(solutions_df["Vehicles"])

mean_distance_best_known = np.mean(best_known_df["Distance"])
mean_distance_solution = np.mean(solutions_df["Distance"])

# Mean absolute error (MAE)
mae_vehicles = np.mean(np.abs(solutions_df["Vehicles"] - best_known_df["Vehicles"]))
mae_distances = np.mean(np.abs(solutions_df["Distance"] - best_known_df["Distance"]))

# Mean relative error (MRE)
mre_vehicles = np.mean(np.abs((solutions_df["Vehicles"] - best_known_df["Vehicles"]) / best_known_df["Vehicles"]))
mre_distances = np.mean(np.abs((solutions_df["Distance"] - best_known_df["Distance"]) / best_known_df["Distance"]))

# Mean ratio 
mr_vehicles = mean_vehicles_solution / mean_vehicles_best_known
mr_distances = mean_distance_solution / mean_distance_best_known

# Percentage error is calculated as (MAE / Actual value) * 100
percentage_error_vehicles = ((mean_vehicles_solution - mean_vehicles_best_known) / mean_vehicles_best_known) * 100
percentage_error_distances = ((mean_distance_solution - mean_distance_best_known) / mean_distance_best_known) * 100


print(f"Vehicle difference: {sum_vehicles_solution - sum_vehicles_best_known}")
print(f"Mean Absolute Error (distance): {mae_distances:.4f}")
print(f"Mean Relative Error (distance): {mre_distances:.4f}")
print(f"Mean Approximation Ratio (distance): {mr_distances:.4f}")
print(f"Percentage Error (vehicles): {percentage_error_vehicles:.2f}%")
print(f"Percentage Error (distances): {percentage_error_distances:.2f}%")

Vehicle difference: 7
Mean Absolute Error (distance): 26.8283
Mean Relative Error (distance): 0.0231
Mean Approximation Ratio (distance): 1.0033
Percentage Error (vehicles): 4.86%
Percentage Error (distances): 0.33%


5000 iterations
- Vehicle difference: 7
- Mean Absolute Error (distance): 26.8283
- Mean Relative Error (distance): 0.0231
- Mean Approximation Ratio (distance): 1.0033
- Percentage Error (vehicles): 4.86%
- Percentage Error (distances): 0.33%

10000 iterations
- Vehicle difference: 6
- Mean Absolute Error (distance): 21.3128
- Mean Relative Error (distance): 0.0188
- Mean Approximation Ratio (distance): 0.9985
- Percentage Error (vehicles): 4.17%
- Percentage Error (distances): -0.15%

negative % error due to some solution having more routes but less travel distances
